src/sdk/pynni/nni/regularized_evolution_tuner/regularized_evolution_tuner.py

```diff
@@ -0,0 +1,169 @@
+                sample.append(random.choice(list(self.population)))
+
+            candidate = max(sample, key=lambda x: x.result)
+            arch = self._mutate_model(candidate)
+            self.history[parameter_id] = arch
+            return arch
+        else:
+            raise nni.NoMoreTrialError
+
+    def receive_trial_result(self, parameter_id, parameters, value, **kwargs):
+        """
+        Record the result from a trial
+
+        Parameters
+        ----------
+        parameter_id : int
+        parameters : dict
+        value : dict/float
+            if value is dict, it should have "default" key.
+            value is final metrics of the trial.
+        """
+        reward = extract_scalar_reward(value)
+        if parameter_id not in self.history:
+            raise RuntimeError('Received parameter_id not in total_data.')
+        params = self.history[parameter_id]
+
+        if self.optimize_mode == OptimizeMode.Minimize:
+            reward = -reward
+
+        self.population.append(FinishedIndividual(parameter_id, params, reward))
+        if len(self.population) > self.population_size:
+            self.population.popleft()
+
+    def update_search_space(self, search_space):
+        """
+        Update search space.
+        Search_space contains the information that user pre-defined.
+
+        Parameters
+        ----------
+        search_space : dict
+        """
+        logger.info('update search space %s', search_space)
+        assert self.search_space is None
+        self.search_space = search_space
+
+        for _, val in search_space.items():
+            if val['_type'] != 'layer_choice' and val['_type'] != 'input_choice':
+                raise ValueError('Unsupported search space type: %s' % (val['_type']))
+
+        self._generate_initial_population()
+
+    def _random_model(self):
+        individual = {}
+        for key, val in self.search_space.items():
+            if val['_type'] == 'layer_choice':
+                idx = random.randint(0, len(val['_value']) - 1)
+                individual[key] = {'_value': val['_value'][idx], '_idx': idx}
+            elif val['_type'] == 'input_choice':
+                candidates = val['_value']['candidates']
+                n_chosen = val['_value']['n_chosen']
+                if n_chosen == None:
+                    raise RuntimeError('Key n_chosen must be set in InputChoice.')
+                idxs = [random.randint(0, len(candidates) - 1) for _ in range(n_chosen)]
+                vals = [candidates[k] for k in idxs]
+                individual[key] = {'_value': vals, '_idx': idxs}
+        return individual
+
+    def _mutate_model(self, model):
+        new_individual = copy.deepcopy(model.parameters)
+        mutate_key = random.choice(list(new_individual.keys()))
+        mutate_val = self.search_space[mutate_key]
+        if mutate_val['_type'] == 'layer_choice':
+            idx = random.randint(0, len(mutate_val['_value']) - 1)
```

Comment:
Suggest to extract this logic into a function, because `_mutate_model` and `_random_model` are essentially the same.